jooq/src/main/java/io/micronaut/configuration/jooq/graal/DefaultReflectionServiceSubstitute.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.configuration.jooq.graal;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+import org.simpleflatmapper.reflect.DefaultReflectionService;
+import org.simpleflatmapper.reflect.asm.AsmFactory;
+
+/**
+ * Disable not supported ASM bytecode manipulation in GraalVM native image runtime.
+ *
+ * @author Lukas Moravec
+ * @since 2.3.5
+ */
+@TargetClass(DefaultReflectionService.class)
+final class DefaultReflectionServiceSubstitute {
```

Comment:
I'm not sure this will work if simple flat mapper is not on the classpath

Reply:
Do you think it will be better to make dedicated submodule? We were discussing this with @ilopmar in https://github.com/micronaut-graal-tests/micronaut-jooq-graal/issues/1#issuecomment-647982231 but every point has its downsides. What do you think?

Reply:
Yes, it fails :-(

I thought that was going to work because `@TargetClass` would be ignored if the class doesn't exist, my bad...

I did a quick test creating a substitution like this one but from another class from Lettuce, because our test application already uses SimpleFlatMapper. This is the error:

```
Fatal error:java.lang.TypeNotPresentException: Type io.lettuce.core.ConnectionBuilder not present
	at sun.reflect.annotation.TypeNotPresentExceptionProxy.generateException(TypeNotPresentExceptionProxy.java:46)
	at sun.reflect.annotation.AnnotationInvocationHandler.invoke(AnnotationInvocationHandler.java:84)
	at com.sun.proxy.$Proxy25.value(Unknown Source)
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.findTargetClass(AnnotationSubstitutionProcessor.java:797)
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.handleClass(AnnotationSubstitutionProcessor.java:252)
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.init(AnnotationSubstitutionProcessor.java:230)
	at com.oracle.svm.hosted.NativeImageGenerator.createDeclarativeSubstitutionProcessor(NativeImageGenerator.java:904)
	at com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:841)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:538)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:453)
	at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.lang.ClassNotFoundException: io.lettuce.core.ConnectionBuilder
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at sun.reflect.generics.factory.CoreReflectionFactory.makeNamedType(CoreReflectionFactory.java:114)
	at sun.reflect.generics.visitor.Reifier.visitClassTypeSignature(Reifier.java:125)
	at sun.reflect.generics.tree.ClassTypeSignature.accept(ClassTypeSignature.java:49)
	at sun.reflect.annotation.AnnotationParser.parseSig(AnnotationParser.java:439)
	at sun.reflect.annotation.AnnotationParser.parseClassValue(AnnotationParser.java:420)
	at sun.reflect.annotation.AnnotationParser.parseMemberValue(AnnotationParser.java:349)
	at sun.reflect.annotation.AnnotationParser.parseAnnotation2(AnnotationParser.java:286)
	at sun.reflect.annotation.AnnotationParser.parseAnnotations2(AnnotationParser.java:120)
	at sun.reflect.annotation.AnnotationParser.parseAnnotations(AnnotationParser.java:72)
	at java.lang.Class.createAnnotationData(Class.java:3521)
	at java.lang.Class.annotationData(Class.java:3510)
	at java.lang.Class.getAnnotations(Class.java:3446)
	at com.oracle.svm.hosted.ImageClassLoader.canLoadAnnotations(ImageClassLoader.java:252)
	at com.oracle.svm.hosted.ImageClassLoader.handleClass(ImageClassLoader.java:371)
	at com.oracle.svm.hosted.ImageClassLoader.handleClassFileName(ImageClassLoader.java:295)
	at com.oracle.svm.hosted.ImageClassLoader.access$000(ImageClassLoader.java:73)
	at com.oracle.svm.hosted.ImageClassLoader$1.lambda$visitFile$0(ImageClassLoader.java:318)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
```

@morki Then I think you need to explore the `onlyWith` we discussed in https://github.com/micronaut-graal-tests/micronaut-jooq-graal/issues/1#issuecomment-647982231

Reply:
@ilopmar and what do you think about dedicated submodule? It feels to me not that hacky :)

Reply:
I think like `io.micronaut.sql:micronaut-sfm` and with `api` scoped dependency to SimpleFlatMapper?

Reply:
I've asked in the GraalVM slack about other alternatives to make that substitution conditional.